luda-editor/server/src/services/image/mod.rs

```diff
@@ -59,104 +96,21 @@ impl rpc::ImageService<SessionDocument> for ImageService {
-                images: objects
+                images: documents
                     .into_iter()
-                    .map(|object| UrlWithLabels {
-                        url: object.url,
-                        labels: s3_key_to_label_list(&object.key, &prefix),
+                    .map(|document| ImageWithLabels {
+                        id: document.image_id,
+                        url: document.get_image_url(),
+                        labels: document.labels,
                     })
                     .collect(),
             })
         })
     }
 }
-
-fn s3_key_to_label_list(key: &str, prefix: &str) -> Vec<Label> {
-    let key_without_prefix = key.strip_prefix(prefix).unwrap();
-    key_without_prefix
-        .split('/')
-        .map(|label_string| {
-            let mut splitted = label_string.split(LABEL_CONNECTOR);
-            let key = splitted.next().unwrap().to_string();
-            let value = splitted.next().unwrap().to_string();
-
-            Label { key, value }
-        })
-        .collect()
-}
-
-fn images_prefix(project_id: rpc::Uuid) -> String {
-    format!("projects/{}/images/", project_id)
-}
-
-fn label_list_to_s3_key(
-    project_id: rpc::Uuid,
-    label_list: &Vec<Label>,
-) -> Result<String, LabelError> {
-    for label in label_list {
-        check_label_component_validity(&label.key)?;
-        check_label_component_validity(&label.value)?;
-    }
-
-    let key = label_list
-        .iter()
-        .map(|label| format!("{}{LABEL_CONNECTOR}{}", label.key, label.value))
-        .collect::<Vec<_>>()
-        .join(LABEL_JOINER);
-
-    // by s3 spec, the key must be less than 1024 bytes
-    if key.len() > 1024 {
-        return Err(LabelError::TooLong);
-    }
-
-    Ok(format!("{prefix}{key}", prefix = images_prefix(project_id)))
-}
-
-const LABEL_CONNECTOR: &str = ":";
-const LABEL_JOINER: &str = "∷";
-
-fn check_label_component_validity(component: &String) -> Result<(), LabelError> {
-    // # Not available by s3
-    // Backslash ("\")
-    // Caret ("^")
-    // Grave accent / back tick ("`")
-    // 'Greater Than' symbol (">")
-    // 'Less Than' symbol ("<")
-    // Left curly brace ("{")
-    // Right curly brace ("}")
-    // Right square bracket ("]")
-    // Left square bracket ("[")
-    // 'Pound' character ("#")
-    // Non-printable ASCII characters (128–255 decimal characters)
-    // Percent character ("%")
-    // Quotation marks
-    // Tilde ("~")
-    // Vertical bar / pipe ("|")
-
-    // # Not available by us
-    // Colon (":")
-    // Proportion ("∷")
-
-    [
-        "\\", "^", "`", ">", "<", "{", "}", "]", "[", "#", "%", "\"", "~", "|", ":", "∷",
-    ]
-    .into_iter()
-    .find(|c| component.contains(c))
-    .map(|c| Err(LabelError::InvalidCharacter(c.to_string())))
-    .unwrap_or(Ok(()))
-}
-
-#[derive(Debug)]
-enum LabelError {
-    InvalidCharacter(String),
-    TooLong,
-}
-crate::simple_error_impl!(LabelError);
```

Comment:
I decide not to use s3 directly for image. I save image metadata(id, labels) to dynamodb and store image binary data to s3.